OpenRA.Mods.Common/Activities/RemoteDockSequence.cs

```diff
@@ -0,0 +1,64 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Traits;
+
+namespace OpenRA.Mods.Common.Activities
+{
+	public class RemoteDockSequence : Activity
+	{
+		protected readonly DockClientManager DockClient;
+		protected readonly Actor DockHostActor;
+		protected readonly IDockHost DockHost;
+		protected bool dockStarted = false;
+		protected readonly int CloseEnough;
+
+		public RemoteDockSequence(DockClientManager client, Actor hostActor, IDockHost host, int closeEnough)
+		{
+			DockClient = client;
+			DockHost = host;
+			DockHostActor = hostActor;
+			CloseEnough = closeEnough;
+			QueueChild(new Wait(host.DockWait));
+		}
+
+		protected override void OnFirstRun(Actor self)
+		{
+			if (!IsCanceling && !DockHostActor.IsDead && DockHostActor.IsInWorld
+				&& (self.CenterPosition - DockHost.DockPosition).Length <= CloseEnough
+				&& DockClient.CanDockAt(DockHostActor, DockHost, false, true))
+			{
+				DockHost.OnDockStarted(DockHostActor, self, DockClient);
+				dockStarted = true;
+			}
+			else
+			{
+				DockClient.UnreserveHost();
+			}
+		}
+
+		public override bool Tick(Actor self)
+		{
+			if (!dockStarted)
+				return true;
+
+			if (!IsCanceling && !DockHostActor.IsDead && DockHostActor.IsInWorld
+				&& (self.CenterPosition - DockHost.DockPosition).Length <= CloseEnough
+				&& !DockClient.OnDockTick(self, DockHostActor, DockHost))
+			{
+				return false;
```

Comment:
unreserve is called inside `OnDockCompleted`, perhaps that should be documented

Reply:
Then it can always run in `OnLastRun` regardless of success or not. (Unreserving twice shouldn't matter.)

Reply:
I think the current implementation is cleaner

Reply:
With the new helper method it indeed looks clean enough. That opens one more question though: Should it also call OnDock*Completed* when docking fails due to cancelling? (I assume you made it this way because cancelling also counts as being completed?)

Reply:
The logic is, if docking was started, it must be completed. It's only not fully true in cases of actor death as it doesn't really matter for the dead actor